### ISSUE TITLE
S3 event with existing true fails if separate stacks have same service/stage value

### ISSUE THREAD

**Issue by et304383** [serverless/serverless#7502]
# Bug Report

## Description
Serverless doesn't seem to handle the custom resource naming correctly when using S3 events across multiple stacks with the same service name (but different stack name values).

**What did you do?**
I tried to add an event on a Lambda function attached to an existing S3 bucket.  At the same time, an existing serverless deployment with the same service name but a different (custom) stack name also has a Lambda function with an S3 event attached.

I am not following an anti-pattern as separating serverless stacks to prevent hitting the 200 resource limit is recommended by the serverless documentation.

**What happened?**
I received an error:
An error occurred: CustomDashresourceDashexistingDashs3LambdaFunction - myservice-dev-custom-resource-existing-s3 already exists in stack arn:aws:cloudformation:us-east-1:999999999999:stack/myservice-clam-av-dev/uuid-goes-here.

**What should've happened?**
Serverless should not be naming the function and this deployment should have went through without naming collision on the function.

**What's the content of your `serverless.yml` file?**
Key area:

```
    events:
      - s3:
          bucket: !Ref CodeMirrorBucket
          event: s3:ObjectCreated:*
          rules:
            - prefix: /
            - suffix: tag
          existing: true
```

What's the output you get when you use the `SLS_DEBUG=*` environment variable (e.g. `SLS_DEBUG=* serverless deploy`)
This isn't relevant.

TLDR: serverless should not give custom names to the custom resource Lambdas it creates as they are going to collide across stacks.

**Comment by medikoo** [comment]
> when using S3 events across multiple stacks 

@et304383 thanks for report.  If I understand correctly, issue is that for two different services which share _service name_ but have different (through customization) _stack name_, function with same name is created to handle existing S3 bucket via custom resources.

I can confirm that happening here: https://github.com/serverless/serverless/blob/80f1c798a293d99529f9e7200227aa0f96ace219/lib/plugins/aws/customResources/index.js#L34

I understand that on your side you want to treat it those two stacks as part of a _same_ service. Still from Framework point of view, each _stack_ is a different service, and overriding the stack name, doesn't change that (it's just instruction to use different name in CloudFormation realm). 

Trying to use same service name, for two different services is likely to introduce issues. Ideally to recover just use different service names.

**Comment by et304383** [comment]
I agree this could be alleviated by using different service names, but there's no value added by having serverless generate these custom function resources with a given name.  It should just rely on the CFN stack generated name which would be different since the stack names are different.

**Comment by medikoo** [comment]
@et304383 in Serverless Framework, it's `service` that's intended to be a unique name for service, which is also used to ensure unique resource names.

`stackName` is just an optional customization for CloudFormation service, it's not intended to be used anywhere else (for resource naming etc.)

**Comment by et304383** [comment]
@medikoo  I'm not suggesting using the variable for anything.  I'm just suggesting that you don't need to specify a function name when generating the CFN resource for the Lambda function backing this custom resource.  If you simply omitted it when generating this resource, everything would be fine.

There's no value to be gained by giving this a specific name, when it's a function no one cares about given it's not part of the application code.

**Comment by medikoo** [comment]
>  I'm just suggesting that you don't need to specify a function name when generating the CFN resource for the Lambda function backing this custom resource.

@et304383 can you clarify? At this point we don't expect anyone to specify this function name. It's automatically generated

**Comment by et304383** [comment]
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-functionname

You shouldn't specify this for the custom Lambda.  If you simply omitted this, CFN would generate the name, and it would be unique across stacks.

**Comment by medikoo** [comment]
@et304383 thanks for clarification. Functions in serverless are named intentionally, so then in a console it's easy to assume which one belongs to which service.

I'm not sure what would be the names created by CloudFormation, but I assume it'll be something more noisy, that'll be hard to coin without a context.

**Comment by et304383** [comment]
@medikoo but no one cares about the name of these functions that only serve to make deployment function.  They are not part of the application.

In fact, no functions in serverless are given explicit names unless you set the name variable:

https://serverless.com/framework/docs/providers/aws/guide/functions#configuration

Which then maps to the FunctionName parameter in CloudFormation.

If you don't specify this value, CFN names the function based on a concatenation of the stack name, the resource name, and a random string (trimmed to 64 chars of course).

For example, if you dropped specifying the name when creating the custom function, it should end up something like:

```
myservice-dev-stackA-CustomDashresourceDashexistingD-AB82MNQ48M
```

In another stack, it might be:
```
myservice-dev-stackB-CustomDashresourceDashexistingD-JMSDF834AP
```

Again, this function is not something the user cares about as it only serves to get around CloudFormation issues and is not related to application functionality.  So by naming it, you're only making scenarios such as mine more difficult while providing no additional value.

**Comment by medikoo** [comment]
> In fact, no functions in serverless are given explicit names unless you set the name variable:
> https://serverless.com/framework/docs/providers/aws/guide/functions#configuration

We do not expect users to explicitly give names, but Framework always configures names for a functions in CF templates (and it's same with this custom resource functions)

> Again, this function is not something the user cares about

Users often care about what resources are created, and they happen complain if they see resources they did not asked for explicitly.
Providing a meaningful names for such resources helps to understand the purpose of a function, and allows us to quickly answer why given function was created, when user questions it.

**Comment by et304383** [comment]
> Users often care about what resources are created, and they happen complain if they see resources they did not ask for explicitly.
> Providing a meaningful names to such resources helps to understand the purpose of a function, and allows us to quickly answer why given function was created, when user questions it.
Then why not use `Ref: AWS::StackName` to prefix it instead of service name?  If the user doesn't specify a custom stack name, the behaviour is identical to what you're doing today.  If the user does specify a stack name (in my case) then collision is avoided.

**Comment by medikoo** [comment]
> Then why not use Ref: AWS::StackName to prefix it instead of service name? 

@et304383 it's service name and not eventual custom stack name, that identifies service and groups its resources.

Changing that, will make a significant breaking change to thea framework.

Again, problem on your side, is that you assumed that custom stack name identifies the service and not a service name.
